(#2597) Add Pester tests to test the package built

corbob opened the pull request:
## Description Of Changes

This PR brings out tests and packages from Test Kitchen into the GitHub repository. This will allow a developer to build the software locally.

## Motivation and Context

Allowing developers to test their changes don't break other established scenarios.

## Testing

At this point all testing has been done on a fairly fresh Windows 10 VM.

1. Install prerequisite PowerShell module Pester: `choco install pester --version 5.3.1`
2. Build the Chocolatey source code using the `./build.bat` script.
3. In an elevated PowerShell prompt, run `./Invoke-Tests.ps1`
4. Ensure that all tests are passing. - There is a known issue (that will get an issue opened for it) Where the test `choco install.Force Installing a Package that is already installed (with a read/delete locked file).Should not have been able to delete the rollback` fails under Windows 10/11, but not under Server 2016 or 2019. All other tests are passing, and this appears to be something with the way the client OS handles things different from a server OS.

Another way this can be tested is with the Vagrantfile put in the `tests` directory. From the `tests` directory, if you run `vagrant up` it will stand up a VM, install the necessary software, then build and test Chocolatey (it will also test it as though it is on Team City which means all of the signing tests will fail)

## Change Types Made

* [ ] Bug fix (non-breaking change)
* [x] Feature / Enhancement (non-breaking change)
* [ ] Breaking change (fix or feature that could cause existing functionality to change)
* [x] PowerShell code changes. (Pester tests were added).

## Related Issue

Fixes #2597 

## Change Checklist

* [ ] Requires a change to the documentation
* [ ] Documentation has been updated
* [ ] Tests to cover my changes, have been added
* [ ] All new and existing tests passed.
* [ ] ~~PowerShell v2 compatibility checked.~~ PowerShell v2 compatibility has not been checked as these tests require Pester 5.3.1 which requires a minimum of PowerShell v3. All tests have been tested in a Windows 10 environment with PowerShell 5.1

vexx32 commented on TESTING.md:
```diff
@@ -0,0 +1,19 @@
+
+### Pester Tests
+
+The Pester tests have been modelled in a way to be testable without installing Chocolatey to your local system. We have made efforts to prevent installing software with the tests, or at least uninstalling any that is installed. That being said, we also provide a `Vagrantfile` with a base configuration that will build the Chocolatey package (inside the VM) and run the tests for you.
```
We should add an example of how we'd recommend to run the tests on the local system here as well, I think. (Even if it's just `Invoke-Pester tests` or something.)

corbob replied:
I've updated the document to indicate you can also use the `./Invoke-Tests.ps1` script, as well as including the parameters you can give it.

vexx32 commented on tests/chocolatey-tests/choco-install.Tests.ps1:
```diff
@@ -0,0 +1,1378 @@
+            $XML.package.metadata.version | Should -Be "1.0.0"
+        }
+
+        # This behaviour was fixed in 0.10.16
+        It "Should not have been able to delete the rollback" -Skip:(-not (Test-ChocolateyVersionEqualOrHigherThan "0.10.16-beta")) {
+            "$env:ChocolateyInstall\lib-bkp\$PackageUnderTest" | Should -Exist
+        }
```
Maybe we should remove this test if it's since been fixed and replace it with a test that it _should_ have been able to delete it?

corbob replied:
That's the thing, if I'm reading the Skip and the comment correctly, this is indicating that the test is to verify that it was fixed.

vexx32 replied:
Ahh okay, I think I see.

vexx32 commented on tests/chocolatey-tests/choco-install.Tests.ps1:
```diff
@@ -0,0 +1,1378 @@
+        }
+    }
+
+    Context "Force Installing a Package that is already installed (with a read/delete locked file)" {
```
The wording of this test implies that the file is locked for the purposes of reading and deletion, but the test themselves appear to be verifying that choco _is_ in fact able to delete the file, because it was open with the FileShare modes that permit deletion?

Maybe something like:
```suggestion
    Context "Force Installing a Package that is already installed (with a locked file that permits read/delete)" {
```

corbob replied:
So this is the part that is thoroughly confusing to me. Reading the current documentation, the suggested change is what the documentation says it does. But looking at the behaviour difference between .Net 4.7.2 and 4.8 (#2690) it seems that the current documentation although it'll tell you is for 4.7.2, might actually be for 4.8...

vexx32 replied:
It might just be that we're relying somewhere on a default behaviour for a file delete operation that changed in the underlying .NET assemblies after 4.7.2, but yeah I agree it's not clear from the documentation.

vexx32 reviewed (changes requested):
A couple small things I think we should fix up here, but the rest of this looks great, solid work! 💖

corbob commented on Invoke-Tests.ps1:
```diff
@@ -0,0 +1,111 @@
+#>
+param(
+  # Path to place Chocolatey test related artifacts.
+  [string]
```
Turns out we had a `.editorconfig` file set to 2 spaces for this. I've updated it to 4 and will run the formatter across all of the added files so we're somewhat consistent.

ferventcoder commented on TESTING.md:
```diff
@@ -0,0 +1,19 @@
+
+### NUnit Tests
+
+To run the NUnit tests get run automtically when you run `./build.bat`, you can also run them without building the entire solution by running `./test.bat`.
```
I think that test.bat still compiles the code - https://github.com/chocolatey/choco/blob/develop/test.bat#L18. I get that you are saying it's not running the entire suite of the build, perhaps change the wording a bit?

corbob replied:
Good point @ferventcoder, that's exactly what I meant by that. I've reworded it to hopefully make it clear that `./test.bat` doesn't complete the full build process.

corbob commented:
@vexx32 I believe I've addressed your reviews, aside from the two regarding locked files. I have switched it back to Draft so we don't accidentally merge it without first rebasing the commits due to some being put in with the intention of squashing into other existing commits.

vexx32 reviewed (approved):
Looks good to me, great work!
